src/metabase/models/revision/diff.clj

```diff
@@ -1,50 +1,76 @@
 
-(defn- diff-string* [k v1 v2]
+(defn- diff-strings* [k v1 v2 identifier]
   (match [k v1 v2]
     [:name _ _]
-    (format "renamed it from \"%s\" to \"%s\"" v1 v2)
+    (deferred-tru "renamed {0} from \"{1}\" to \"{2}\"" identifier v1 v2)
+
+    [:description nil _]
+    (deferred-tru "added a description")
+
+    [:description (_ :guard some?) _]
+    (deferred-tru "changed the description")
 
     [:private true false]
-    "made it public"
+    (deferred-tru "made {0} public" identifier)
 
     [:private false true]
-    "made it private"
+    (deferred-tru "made {0} private" identifier)
+
+    [:archived false true]
+    (deferred-tru "unarchived this")
+
+    [:archived true false]
+    (deferred-tru "archived this")
+
+    [:dataset false true]
+    (deferred-tru "turned this into a model")
 
-    [:updated_at _ _]
-    nil
+    [:dataset false false]
+    (deferred-tru "changed this from a model to a saved question")
+
+    [:display _ _]
+    (deferred-tru "changed the display from {0} to {1}" v1 v2)
 
     [:result_metadata _ _]
-    nil
+    (deferred-tru "edited the metadata")
 
     [:dataset_query _ _]
-    "modified the query"
+    (deferred-tru "modified the query")
+
+    [:collection_id nil (coll-id :guard int?)]
+    (deferred-tru "moved {0} to {1}" identifier (t2/select-one-fn :name 'Collection coll-id))
+
+    [:collection_id (prev-coll-id :guard int?) (coll-id :guard int?)]
+    (deferred-tru "moved {0} from {1} to {2}"
+      identifier
+      (t2/select-one-fn :name 'Collection prev-coll-id)
+      (t2/select-one-fn :name 'Collection coll-id))
 
     [:visualization_settings _ _]
-    "changed the visualization settings"
+    (deferred-tru "changed the visualization settings")
 
-    [_ _ _]
-    (format "changed %s from \"%s\" to \"%s\"" (name k) v1 v2)))
+    :else nil))
 
 (defn build-sentence
   "Join parts of a sentence together to build a compound one."
   [parts]
   (when (seq parts)
     (cond
       (= (count parts) 1) (str (first parts) \.)
-      (= (count parts) 2) (format "%s and %s." (first parts) (second parts))
-      :else               (format "%s, %s" (first parts) (build-sentence (rest parts))))))
-
-(defn diff-string
-  "Create a string describing how `o1` is different from `o2`.
-   The directionality of the statement should indicate that `o1` changed into `o2`."
-  [t before after]
-  (when before
-    (let [ks (keys before)]
-      (some-> (filter identity (for [k ks]
-                                 (diff-string* k (k before) (k after))))
-              build-sentence
-              (str/replace-first #" it " (format " this %s " t))))))
+      (= (count parts) 2) (str (first parts) " " (deferred-tru "and")  " " (second parts) \.)
+      :else               (str (first parts) ", " (build-sentence (rest parts))))))
+
+(defn diff-strings
+  "Create a seq of string describing how `o1` is different from `o2`.
+  The directionality of the statement should indicate that `o1` changed into `o2`."
```

Comment:
This docstring is confusing because `o1` and `o2` refer to the arguments of `diff-strs`, not this function.

A suggestion: any reader of `diff-strings` needs to understand that `before` and `after` are results from `data/diff`. This could be avoided if (1) `diff-strings` was inlined in `diff-strs` `:default` body, or (2) we refactored so `data/diff` was called inside `diff-strings` instead. Then `diff-strs` `:default` body would simply be:
```
(defmethod diff-strs :default
  [model o1 o2]
  (when (not= o1 o2)
    (diff-strings (name model) o1 o2)))
```
But then at that point you might as well inline `diff-strings`, because this function is not doing much on its own.

Reply:
I moved the diff into `diff-strings*` 4f6f0bd02e